### ISSUE TITLE
International Track & Field 2000 won't start (TLB Exception)

### ISSUE THREAD

**Issue by loganmc10** [mupen64plus/mupen64plus-core#311]
Game doesn't boot, this is printed over and over:

```
tlb exception !!! @ 20, 1, add:80021ec4
```

**Comment by loganmc10** [comment]
I have a very similar issue with Midway's Greatest Arcade Hits Vol. 1 (won't boot):

```
tlb exception !!! @ 20, 1, add:80051714
```

I wonder if this is related to https://github.com/mupen64plus/mupen64plus-core/issues/221. They all have an ```add`` value of over 0x80000000

**Comment by loganmc10** [comment]
Both of these games work with CEN64, so I assume it's not the same problem as #221

**Comment by loganmc10** [comment]
Here is the info for the TLB entry that is written right before this starts happening (these numbers are in hex):


![image](https://user-images.githubusercontent.com/848146/27146246-e7029f92-50f5-11e7-9023-7071f7dd95b0.png)

**Comment by fzurita** [comment]
Sounds like a similar issue to what I was having in Goldeneye when loading save states. The TLB mapping was giving a completely bogus memory address.

**Comment by loganmc10** [comment]
Yeah, I don't know if it's the TLB or something else. I got rid of the current TLB implementation and replaced it with CEN64's (it's not terribly hard since the code structure is really similar), other games that use TLB like Conkers/Mario 64/Golden Eye still worked, but this game still didn't.

**Comment by loganmc10** [comment]
Potentially a similar issue in Rat Attack!, which also doesn't start, however this is only printed once:

```
tlb exception !!! @ 0, 1, add:8009b8a8
```

**Comment by loganmc10** [comment]
I believe I've located the condition that is unique to these games. The MIPS "load instructions" (inside https://github.com/mupen64plus/mupen64plus-core/blob/master/src/device/r4300/mips_instructions.def) have the conditional statement:

```
if (*r4300_address(r4300)) {
```
That statement is normally true, but in these games, it evaluates to false once, that is where the game locks up, specifically in the LB and LW instructions. I'm not sure what to do in the "else" case, but it seems like they aren't being handled properly.

**Comment by bsmiles32** [comment]
This check on address is "misleading". It checks if address translation succeded or not. This is a "legacy" of liberal use of global variables.
I think at some point we should rework the TLB implementation. Maybe after the big memory refactorings.

**Comment by loganmc10** [comment]
Yeah I just noticed that only these 2 games (International Track & Field and Midway Arcade) ever trigger a "false" result on that check. Even other games that seem to generate TLB misses like Conkers (if you uncomment https://github.com/mupen64plus/mupen64plus-core/blob/master/src/device/r4300/tlb.c#L142 alot gets printed out in Conkers), don't trigger a "false" result on this check.

I don't really know what it means, I just noticed that it seems to be unique to these games as far as I can tell

**Comment by bsmiles32** [comment]
There is also the possibility that the translated address is indeed 0, but as we use this value as a TLB error sentinel, we go to the wrong branch.

**Comment by loganmc10** [comment]
I believe (I could be wrong) that the issue is with the SW/LW instructions. It seems that there are cases when these instructions can write/read to the ROM, instead of memory?

Project64 handles these cases like this:
https://github.com/project64/project64/blob/master/Source/Project64-core/N64System/Mips/MemoryVirtualMem.cpp#L469-L477
https://github.com/project64/project64/blob/master/Source/Project64-core/N64System/Mips/MemoryVirtualMem.cpp#L798-L803

Again, these 2 games are the only games I see that trigger these checks when I port them over to mupen64plus. I haven't figured out how to get it to write/read the correct address however.

The comments in the 1964 source code have a related message:
> in LW/SW opcode, you need to pass me the real value so I can write the real value into the IO register and I can trigger correct hardware functions Remember, in SW/LW opcode, you still need to write the real value to the memory locate because the SW/LW may be accessing non io registers (at most time)

The English is a little bit broken and the source code is hard to follow, so I'm not exactly sure how 1964 got this working.
